src/dbus/methods/room_get.c

```diff
@@ -0,0 +1,269 @@
+{
+    GVariantBuilder *core_builder;
+    GVariant *core_dict_array;
+
+    core_builder = g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+
+    create_gr_sync(&core->base, core_builder);
+    g_variant_builder_add(core_builder, "{sv}", "room_name", g_variant_new_string(core->room_name));
+    g_variant_builder_add(core_builder, "{sv}", "teams_switched", g_variant_new_int32(core->teams_switched));
+    g_variant_builder_add(core_builder, "{sv}", "private", g_variant_new_int32(core->private));
+    g_variant_builder_add(core_builder, "{sv}", "can_start", g_variant_new_int32(core->can_start));
+    g_variant_builder_add(core_builder, "{sv}", "team_balanced", g_variant_new_int32(core->team_balanced));
+    g_variant_builder_add(core_builder, "{sv}", "min_ready_players", g_variant_new_int32(core->min_ready_players));
+    g_variant_builder_add(core_builder, "{sv}", "num_players", g_variant_new_int32(core->num_players));
+
+    core_dict_array = g_variant_builder_end(core_builder);
+
+    g_variant_builder_add(builder, "{sv}", "core", core_dict_array);
+    g_variant_builder_unref(core_builder);
+}
+
+static void rlist_to_array(struct gameroom *rl, GVariantBuilder *builder)
+{
+    GVariantBuilder *item_builder;
+
+    item_builder = g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+
+    g_variant_builder_add(item_builder, "{sv}", "room_id", g_variant_new_string(rl->room_id));
+    g_variant_builder_add(item_builder, "{sv}", "room_type", g_variant_new_string(
+                                                                        rl->room_type == ROOM_PVE_PRIVATE ? "PVE_PRIVATE":
+                                                                        rl->room_type == ROOM_PVP_PUBLIC ? "PVP_PUBLIC":
+                                                                        rl->room_type == ROOM_PVP_CLANWAR ? "PVP_CLANWAR":
+                                                                        rl->room_type == ROOM_PVP_QUICKPLAY ? "PVP_QUICKPLAY":
+                                                                        rl->room_type == ROOM_PVE_QUICKPLAY ? "PVE_QUICKPLAY": "PVP_RATING"));
+
+    create_gameroom_core(&rl->core, item_builder);
+    create_gameroom_custom_params(&rl->custom_params, item_builder);
+    create_gameroom_mission(&rl->mission, item_builder);
+    create_gameroom_session(&rl->session, item_builder);
+    create_gameroom_room_master(&rl->room_master, item_builder);
+    create_gameroom_regions(&rl->regions, item_builder);
+    create_gameroom_auto_start(&rl->auto_start, item_builder);
+
+    if(rl->clan_war.clan_1 != NULL || rl->clan_war.clan_2 != NULL)
+        create_gameroom_clan_war(&rl->clan_war, item_builder);
+
+    g_variant_builder_add(builder, "a{sv}", item_builder);
+    g_variant_builder_unref(item_builder);
+}
+
+void cmd_room_get_dbus_cb(struct list *room_list, void *args)
+{
+    struct cb_args *a = (struct cb_args *) args;
+    struct list *rl = room_list;
```

Comment:
I'm not sure why you need this variable, instead of directly using room_list

Also, I noticed (idk why I did that) that the callback has ownership on the list given in parameter. Here we are leaking memory with each call to  room get.. see line 157 of xmpp IQ game room get.

I suggest to do instead remove the else and always free the list  from the module that created it.... or you need to free it in the callback.

Either way, find what suits you the best, I'm not home to check with my part of the code.

Reply:
done